### PR TITLE
fix(upgrade): Update types for TypeScript nullability support

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: packages/upgrade/src/common/downgrade_component.ts [hunk 1]
```diff
@@ -145,10 +145,10 @@ class ParentInjectorPromise {
     this.injector = injector;
 
     // Store the real injector on the element.
-    this.element.data(this.injectorKey, injector);
+    this.element.data !(this.injectorKey, injector);
 
     // Release the element to prevent memory leaks.
-    this.element = null;
+    this.element = null !;
```
Comment (petebacondarwin) [inline packages/upgrade/src/common/downgrade_component.ts]: surely this can't be right??
You are asserting that `null` is strictly not `null`!

**Comment by petebacondarwin** [comment]
What is the motivation behind this change? Is it purely to allow us to turn on strict null checks?
I feel that it has been somewhat automated and that it actually makes the code less maintainable rather than more.
If we are going to do this then we should perhaps do a better refactoring to take into account the benefits of nullability support.

File: packages/upgrade/src/common/downgrade_component.ts (hunk 1 above)
Comment (mhevery) [inline packages/upgrade/src/common/downgrade_component.ts]: This is in clean up code where we release the memory by clearing it. We could say that the `element` field is nullable, but than we would have to put `!` everywhere else. What we are saying here, is that the code is finished and we no longer want to keep this element and no more code will run.

So you have a trade off to make. Make the field nullable and have `!` on every reference, or cheat at cleanup time. I chose the former.

**Review by mhevery — COMMENTED** [review]
The motivation is to turn on `strictNullChecks` so that our customers may turn on `strictNullChecks` in their codebases. You are right some changes are suboptimal, but given the number of changes I had to do sometimes corners got cut.  Will will have to have clean up CLs afterwards.

**Pull request closed without merging**